test/e2e_node/container_lifecycle_test.go

```diff
@@ -5905,7 +5905,233 @@ var _ = SIGDescribe(feature.SidecarContainers, "Containers Lifecycle", func() {
+
+			expectPodTerminationContainerStatuses(pod.Status.ContainerStatuses, map[string]podTerminationContainerStatus{
+				regular1: {exitCode: 0, reason: "Completed"},
+			})
+			expectPodTerminationContainerStatuses(pod.Status.InitContainerStatuses, map[string]podTerminationContainerStatus{
+				restartableInit1: {exitCode: 0, reason: "Completed"},
+				restartableInit2: {exitCode: 137, reason: "Error"},
+				restartableInit3: {exitCode: 0, reason: "Completed"},
+			})
+		})
+	})
+
+	ginkgo.When("A restartable init container with startup probe fails initially", func() {
+		ginkgo.It("should continue probing and allow regular container to start after the restartable init recovers", func(ctx context.Context) {
+			restartableInit := "buggy-restartable-init"
+			regularContainer := "regular-container"
+
+			restartPolicyAlways := v1.ContainerRestartPolicyAlways
+
+			pod := &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "restartable-init-startup-probe-fix",
+				},
+				Spec: v1.PodSpec{
+					RestartPolicy: v1.RestartPolicyNever,
+					Volumes: []v1.Volume{{
+						Name: "workdir",
+						VolumeSource: v1.VolumeSource{
+							EmptyDir: &v1.EmptyDirVolumeSource{},
+						},
+					}},
+					InitContainers: []v1.Container{{
+						Name:          restartableInit,
+						Image:         busyboxImage,
+						RestartPolicy: &restartPolicyAlways,
+						VolumeMounts: []v1.VolumeMount{{
+							Name:      "workdir",
+							MountPath: "/work",
+						}},
+						Command: []string{"sh", "-c", "if [ ! -f /work/first_run_done ]; then echo 'First run: creating marker and exiting with 1'; touch /work/first_run_done; exit 1; else echo 'Second run: marker found, running as sidecar'; sleep 120; fi"},
+						StartupProbe: &v1.Probe{
+							InitialDelaySeconds: 5,
+							PeriodSeconds:       2,
+							FailureThreshold:    10,
+							ProbeHandler: v1.ProbeHandler{
+								Exec: &v1.ExecAction{
+									Command: []string{"/bin/true"},
+								},
+							},
+						},
+					}},
+					Containers: []v1.Container{{
+						Name:  regularContainer,
+						Image: imageutils.GetPauseImageName(),
+						StartupProbe: &v1.Probe{
+							InitialDelaySeconds: 5,
+							PeriodSeconds:       5,
+							ProbeHandler: v1.ProbeHandler{
+								Exec: &v1.ExecAction{
+									Command: []string{"/bin/true"},
+								},
+							},
+						},
+					}},
+				},
+			}
+
+			preparePod(pod)
+			client := e2epod.NewPodClient(f)
+			pod = client.Create(ctx, pod)
+
+			ginkgo.By("Waiting for init container to fail and restart at least once")
+			framework.ExpectNoError(
+				e2epod.WaitForPodCondition(ctx, f.ClientSet, pod.Namespace, pod.Name,
+					"restartable init restarted", 90*time.Second,
+					func(p *v1.Pod) (bool, error) {
+						for _, st := range p.Status.InitContainerStatuses {
+							if st.Name == restartableInit && st.RestartCount > 0 {
+								ginkgo.By(fmt.Sprintf("Init container %s has restarted %d times", restartableInit, st.RestartCount))
```

Comment:
I guess it would be better to use `framework.Logf` here instead of `ginkgo.By` and the others